File: src/lib/clients/content-layout-policy.ts
/**
 * The one place that decides what a folder *means*.
 *
 * Two things flatten torrent layouts: the planner, which rewrites paths before
 * the chunk store is built, and the repair, which lifts folders that are
 * already on disk. If they disagree by even one rule, a torrent added today
 * lands somewhere its own resume data is not, and the bytes are downloaded
 * twice. So both import their rules from here and nowhere else.
 */
import path from "node:path";

/**
 * Structures a player, reader or console looks up by name. Dropping one breaks
 * playback rather than merely moving files, so it is never dropped — not at
 * any depth, not by either half.
 */
export const PROTECTED_FOLDER =
  /^(?:BDMV|VIDEO_TS|AUDIO_TS|CERTIFICATE|PRIVATE|AVCHD|BDROM|STREAM|PLAYLIST|CLIPINF|PS3_GAME|PS3_UPDATE|USRDIR|SYSTEM\.BIN)$/i;

/**
 * Folders that organise content rather than wrap it. `Disc 2` and `Season 02`
 * are the only thing keeping two sets of identically named files apart, so
 * dropping one silently merges them.
 *
 * These are still dropped when the destination path already says the same
 * thing — `Season 01` inside `…/Season 01` is pure repetition.
 */
export const STRUCTURAL_FOLDER =
  /^(?:season[\s._-]*\d{1,3}|s\d{1,3}|series[\s._-]*\d{1,3}|specials?|extras?|featurettes?|bonus|subs?|subtitles?|sample|disc[\s._-]*\d+|cd[\s._-]*\d+|dvd[\s._-]*\d+|dis[ck][\s._-]*\d+|vol(?:ume)?[\s._-]*\d+|part[\s._-]*\d+)$/i;

/** Tokens that describe *which* season, not *which* release. */
const SEASON_MARKER = /^s(\d{1,3})$/;

/**
 * Tokens that describe the encode rather than the work. Two names that share
 * only these share nothing meaningful — `1080p x265` is not a title.
 */
const TECHNICAL_TOKEN =
  /^(?:\d{3,4}p|4k|uhd|hdr|hdr10|sdr|x26[45]|h26[45]|hevc|avc|xvid|divx|web|webrip|webdl|bdrip|brrip|bluray|blu|ray|dvdrip|hdtv|remux|aac|ac3|eac3|dts|dd|ddp|flac|mp3|opus|atmos|truehd|\d+bit|bits|\d+|dual|multi|audio|subs?|repack|proper|internal|complete|completo|batch|pack|extended|final)$/;

/** Reserved characters `fs-chunk-store` strips from every basename it writes. */
const RESERVED_FILENAME = /[<>:"/\\|?*\u0000-\u001F]/g;

/** Splits a torrent path on either separator, dropping empty segments. */
export function segments(p: string): string[] {
  return p
    .split(/[\\/]+/)
    .map((s) => s.trim())
    .filter(Boolean);
}

/**
 * Compares folder names the way a person would: `Season 1`, `Season 01` and
 * `S01` are the same folder. Leading zeros are only stripped off a season or
 * episode marker — `007` and `7` are different films.
 */
export function folderKey(name: string): string {
  return name
    .toLowerCase()
    .replace(/[._]+/g, " ")
    .replace(/^s(?=\d{1,3}\b)/, "season ")
    .replace(/\s+/g, " ")
    .replace(/^(season|series|disc|disk|cd|dvd|vol|volume|part)\s+0*(\d+)/, "$1 $2")
    .trim();
}

/**
 * Tokenises a release name, folding season markers so `S01` and `Season 01`
 * are one token.
 */
function releaseTokens(name: string): Set<string> {
  const flat = name
    .toLowerCase()
    .replace(/\bseason\s*(\d{1,3})\b/g, "s$1")
    .replace(/\bs(\d{1,3})\b/g, (_, d: string) => `s${Number(d)}`)
    .replace(/\be(\d{1,3})\b/g, (_, d: string) => `e${Number(d)}`);
  return new Set(flat.split(/[^a-z0-9]+/).filter(Boolean));
}

/**
 * True when two folder names are the same release wrapped twice.
 *
 * Deliberately not a similarity score. A threshold cannot tell
 * `… Dual Audio` from `… English Audio` (0.75 alike), or `x264` from `x265`
 * (0.67) — and getting that wrong merges two different encodes into one
 * folder. So the test is exact: the names must agree on every token, with two
 * escape hatches and no others.
 *
 *   - A season marker may differ, but **only** if the destination already
 *     names that season. `Frieren Complete` vs `Frieren S01` going into
 *     `Anime/Frieren` is not a duplicate — the inner folder is the only thing
 *     saying which season these files are.
 *   - The names must share at least one token that is not pure encode jargon,
 *     so two unrelated releases cannot match on `1080p x265` alone.
 *
 * False negatives merely leave a folder nested. False positives destroy the
 * only thing keeping two releases apart, so the asymmetry is intentional.
 */
export function isSameRelease(
  a: string,
  b: string,
  destKeys?: Set<string>,
): boolean {
  const ta = releaseTokens(a);
  const tb = releaseTokens(b);
  if (ta.size === 0 || tb.size === 0) return false;

  let sharedMeaningful = 0;
  for (const t of ta) if (tb.has(t) && !TECHNICAL_TOKEN.test(t)) sharedMeaningful++;
  if (sharedMeaningful === 0) return false;

  const allowed = (t: string): boolean => {
    const season = SEASON_MARKER.exec(t);
    if (!season) return false;
    // Only droppable if the destination already says which season this is.
    return destKeys?.has(`season ${Number(season[1])}`) ?? false;
  };

  for (const t of ta) if (!tb.has(t) && !allowed(t)) return false;
  for (const t of tb) if (!ta.has(t) && !allowed(t)) return false;
  return true;
}

/**
 * The destination components a container folder may repeat.
 *
 * Only the tail is considered. Matching every component would let a folder
 * named `Downloads` or `Media` be dropped because the library happens to live
 * under a folder of that name, which says nothing about the torrent.
 */
export function destinationKeys(destPath?: string | null): Set<string> {
  return new Set(segments(destPath ?? "").slice(-3).map(folderKey));
}

/**
 * Decides whether one container level may be dropped. Shared verbatim by the
 * planner and the repair — this function *is* the agreement between them.
 *
 * @param name     the folder in question
 * @param depth    0 for the torrent's own container root
 * @param destKeys from {@link destinationKeys}
 * @param ancestors folders already dropped, outermost first
 */
export function mayDropFolder(
  name: string,
  depth: number,
  destKeys: Set<string>,
  ancestors: readonly string[],
): boolean {
  if (PROTECTED_FOLDER.test(name)) return false;

  const repeatsDestination = destKeys.has(folderKey(name));

  // A structural folder is the only thing separating two discs or two
  // seasons. It goes only when the destination already says the same thing.
  if (STRUCTURAL_FOLDER.test(name)) return repeatsDestination;

  // Depth 0 is the torrent's container root, redundant by construction —
  // this is exactly qBittorrent's `contentLayout=NoSubfolder`, so a download
  // moved between the two clients still finds its files.
  if (depth === 0) return true;

  // Deeper down, absence of meaning is not evidence. Require proof.
  return (
    repeatsDestination ||
    ancestors.some((prior) => isSameRelease(prior, name, destKeys))
  );
}

/** True when the destination path itself already names one season. */
export function destinationNamesSeason(destPath?: string | null): boolean {
  const segs = segments(destPath ?? "");
  const last = segs[segs.length - 1];
  return last !== undefined && /^(?:season|series|s)[\s._-]*\d{1,3}$/i.test(last);
}

/**
 * Renames a release folder that names exactly one season into `Season NN`.
 *
 * A multi-season pack cannot have its season folders *dropped* — they are the
 * only thing keeping two sets of identically numbered episodes apart — so the
 * batch root goes and the library is left holding
 * `Solo Leveling/Solo Leveling S01 1080p … x265-EMBER/`. Renaming gives the
 * same layout a single-season download would have produced.
 *
 * Deliberately narrow, because a wrong rename merges two releases:
 *
 *   - never a protected or already-structural folder (`Season 01`, `Specials`,
 *     `Extras`, `BDMV` …) — those are either correct already or not seasons;
 *   - the name must carry at least one encode token, so a *title* that happens
 *     to contain `S2` is not mistaken for a season folder;
 *   - the name must resolve to exactly one season number, so `S01-S02` and
 *     other ranges are left alone.
 *
 * Returns null to leave the folder untouched, which is always the safe answer.
 */
export function seasonFolderRename(name: string): string | null {
  if (PROTECTED_FOLDER.test(name)) return null;
  if (STRUCTURAL_FOLDER.test(name)) return null;

  const found = [
    ...name.matchAll(/(?:^|[^a-z0-9])s(?:eason)?[\s._-]*(\d{1,3})(?![0-9])/gi),
  ];
  const seasons = new Set(found.map((m) => Number(m[1])));
  if (seasons.size !== 1) return null;

  const tokens = name.toLowerCase().split(/[^a-z0-9]+/).filter(Boolean);
  if (!tokens.some((t) => TECHNICAL_TOKEN.test(t))) return null;

  const season = [...seasons][0];
  return `Season ${String(season).padStart(2, "0")}`;
}

/**
 * The key two paths collide on *as files on disk*, which is not the same as
 * being equal strings.
 *
 * `fs-chunk-store` strips reserved characters from every basename it writes,
 * on every platform, so `Episode 01: Arrival.mkv` and `Episode 01 Arrival.mkv`
 * are one file. Windows adds case insensitivity and ignores trailing dots and
 * spaces. Comparing raw strings misses all of it and lets two files quietly
 * verify over each other.
 */
export function physicalKey(relativePath: string): string {
  const parts = segments(relativePath);
  const cleaned = parts.map((seg, i) => {
    let s = seg;
    // Only the basename is sanitised by the store; directories are joined
    // as-is, but Windows still folds their case and trailing punctuation.
    if (i === parts.length - 1) s = s.replace(RESERVED_FILENAME, "");
    if (process.platform === "win32") s = s.replace(/[. ]+$/, "");
    return s;
  });
  const joined = cleaned.join("/");
  return process.platform === "win32" ? joined.toLowerCase() : joined;
}

/**
 * True when `child` is at or below `root`, resolving `..` first.
 *
 * Compares path segments rather than the string prefix: a real folder called
 * `..stfolder` (Syncthing writes one) starts with `..` without escaping
 * anything.
 */
export function isInside(root: string, child: string): boolean {
  const rel = path.relative(path.resolve(root), path.resolve(child));
  if (rel === "") return true;
  if (path.isAbsolute(rel)) return false;
  return rel !== ".." && !rel.startsWith(`..${path.sep}`);
}

File: server/TorrentFlow.Engine/Layout/ContentLayoutPolicy.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace TorrentFlow.Engine.Layout;

/// <summary>
/// Port of content-layout-policy.ts: the one place that decides what a folder <em>means</em>. The planner and every
/// caller take their rules from here, so a torrent can never be laid out by two disagreeing sets of rules.
/// Regexes run in ECMAScript mode so <c>\d</c>, <c>\w</c> and <c>\b</c> match exactly what the TypeScript matched.
/// </summary>
internal static class ContentLayoutPolicy
{
    private const RegexOptions Js = RegexOptions.ECMAScript | RegexOptions.Compiled;
    private const RegexOptions JsI = Js | RegexOptions.IgnoreCase;

    /// <summary>Structures a player, reader or console looks up by name. Never dropped, at any depth.</summary>
    private static readonly Regex ProtectedFolder = new(
        @"^(?:BDMV|VIDEO_TS|AUDIO_TS|CERTIFICATE|PRIVATE|AVCHD|BDROM|STREAM|PLAYLIST|CLIPINF|PS3_GAME|PS3_UPDATE|USRDIR|SYSTEM\.BIN)$", JsI);

    /// <summary>
    /// Folders that organise content rather than wrap it: <c>Disc 2</c> and <c>Season 02</c> are the only thing keeping
    /// two sets of identically named files apart. Still dropped when the destination already says the same thing.
    /// </summary>
    private static readonly Regex StructuralFolder = new(
        @"^(?:season[\s._-]*\d{1,3}|s\d{1,3}|series[\s._-]*\d{1,3}|specials?|extras?|featurettes?|bonus|subs?|subtitles?|sample|disc[\s._-]*\d+|cd[\s._-]*\d+|dvd[\s._-]*\d+|dis[ck][\s._-]*\d+|vol(?:ume)?[\s._-]*\d+|part[\s._-]*\d+)$", JsI);

    private static readonly Regex SeasonMarker = new(@"^s(\d{1,3})$", Js);

    /// <summary>Tokens that describe the encode rather than the work; <c>1080p x265</c> is not a title.</summary>
    private static readonly Regex TechnicalToken = new(
        @"^(?:\d{3,4}p|4k|uhd|hdr|hdr10|sdr|x26[45]|h26[45]|hevc|avc|xvid|divx|web|webrip|webdl|bdrip|brrip|bluray|blu|ray|dvdrip|hdtv|remux|aac|ac3|eac3|dts|dd|ddp|flac|mp3|opus|atmos|truehd|\d+bit|bits|\d+|dual|multi|audio|subs?|repack|proper|internal|complete|completo|batch|pack|extended|final)$", Js);

    /// <summary>Reserved characters a torrent store strips from every basename it writes.</summary>
    private static readonly Regex ReservedFilename = new("[<>:\"/\\\\|?*\\u0000-\\u001F]", Js);

    private static readonly Regex Separators = new(@"[\\/]+", Js);
    private static readonly Regex DotsUnderscores = new(@"[._]+", Js);
    private static readonly Regex LeadingSeasonS = new(@"^s(?=\d{1,3}\b)", Js);
    private static readonly Regex Whitespace = new(@"\s+", Js);
    private static readonly Regex NumberedFolder = new(@"^(season|series|disc|disk|cd|dvd|vol|volume|part)\s+0*(\d+)", Js);
    private static readonly Regex SeasonWord = new(@"\bseason\s*(\d{1,3})\b", Js);
    private static readonly Regex SeasonToken = new(@"\bs(\d{1,3})\b", Js);
    private static readonly Regex EpisodeToken = new(@"\be(\d{1,3})\b", Js);
    private static readonly Regex NonAlnum = new(@"[^a-z0-9]+", Js);
    private static readonly Regex SeasonSegment = new(@"^(?:season|series|s)[\s._-]*\d{1,3}$", JsI);
    private static readonly Regex SeasonMention = new(@"(?:^|[^a-z0-9])s(?:eason)?[\s._-]*(\d{1,3})(?![0-9])", JsI);
    private static readonly Regex TrailingDotsSpaces = new(@"[. ]+$", Js);

    private static readonly Regex EpisodeMarker = new(@"(?:^|[^a-z0-9])s(\d{1,3})[\s._-]*e(\d{1,4})(?![0-9])", JsI);
    private static readonly Regex CrossEpisodeMarker = new(@"(?:^|[^a-z0-9])(\d{1,2})x(\d{2,3})(?![0-9])", JsI);
    private static readonly Regex SampleName = new(@"(?:^|[^a-z0-9])sample(?:[^a-z0-9]|$)", JsI);
    private static readonly Regex ExtrasFolder = new(
        @"^(?:samples?|extras?|featurettes?|bonus|trailers?|behind[\s._-]*the[\s._-]*scenes|deleted[\s._-]*scenes|interviews?)$", JsI);
    private static readonly Regex SubsFolder = new(@"^(?:subs?|subtitles?)$", JsI);

    private static readonly HashSet<string> VideoExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".mkv", ".mp4", ".m4v", ".avi", ".mov", ".wmv", ".ts", ".m2ts", ".mts", ".webm", ".mpg", ".mpeg", ".flv", ".ogv", ".divx", ".rmvb",
    };

    private static readonly HashSet<string> SubtitleExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".srt", ".ass", ".ssa", ".sub", ".idx", ".vtt", ".sup", ".smi",
    };

    private static string Num(string digits) => int.Parse(digits, CultureInfo.InvariantCulture).ToString(CultureInfo.InvariantCulture);

    /// <summary>
    /// The single episode a name identifies (<c>S01E03</c> or <c>1x03</c>), as <c>s1e3</c>. Null when it names none or
    /// more than one: a folder is only an episode's own folder when there is no doubt which episode.
    /// </summary>
    public static string? EpisodeKey(string name)
    {
        var keys = EpisodeMarker.Matches(name).Concat(CrossEpisodeMarker.Matches(name))
            .Select(m => $"s{Num(m.Groups[1].Value)}e{Num(m.Groups[2].Value)}").ToHashSet(StringComparer.Ordinal);
        return keys.Count == 1 ? keys.First() : null;
    }

    public static bool IsSubsFolder(string name) => SubsFolder.IsMatch(name);

    public static bool IsSeasonFolder(string name) => SeasonSegment.IsMatch(name);

    /// <summary>
    /// What a file is for the layout. A video is the content itself; a sample, or anything under an extras folder, is not,
    /// whatever its extension — it may be moved aside, never given the episode's place.
    /// </summary>
    public static LayoutRole Role(IReadOnlyList<string> segments)
    {
        var name = segments[^1];
        var ext = Path.GetExtension(name);
        if (SubtitleExtensions.Contains(ext)) return LayoutRole.Subtitle;
        if (!VideoExtensions.Contains(ext)) return LayoutRole.Extra;
        if (SampleName.IsMatch(Path.GetFileNameWithoutExtension(name))) return LayoutRole.Extra;
        for (var i = 0; i < segments.Count - 1; i++)
            if (ExtrasFolder.IsMatch(segments[i])) return LayoutRole.Extra;
        return LayoutRole.Video;
    }

    public static bool IsProtected(string name) => ProtectedFolder.IsMatch(name);

    public static bool IsStructural(string name) => StructuralFolder.IsMatch(name);

    /// <summary>Splits a torrent path on either separator, dropping empty segments.</summary>
    public static List<string> Segments(string? p) =>
        Separators.Split(p ?? "").Select(s => s.Trim()).Where(s => s.Length > 0).ToList();

    /// <summary>
    /// Compares folder names the way a person would: <c>Season 1</c>, <c>Season 01</c> and <c>S01</c> are one folder.
    /// Leading zeros are only stripped off a season or episode marker — <c>007</c> and <c>7</c> are different films.
    /// </summary>
    public static string FolderKey(string name)
    {
        var s = name.ToLowerInvariant();
        s = DotsUnderscores.Replace(s, " ");
        s = LeadingSeasonS.Replace(s, "season ", 1);
        s = Whitespace.Replace(s, " ");
        s = NumberedFolder.Replace(s, "$1 $2", 1);
        return s.Trim();
    }

    private static HashSet<string> ReleaseTokens(string name)
    {
        var flat = name.ToLowerInvariant();
        flat = SeasonWord.Replace(flat, "s$1");
        flat = SeasonToken.Replace(flat, m => "s" + Num(m.Groups[1].Value));
        flat = EpisodeToken.Replace(flat, m => "e" + Num(m.Groups[1].Value));
        return NonAlnum.Split(flat).Where(t => t.Length > 0).ToHashSet(StringComparer.Ordinal);
    }

    /// <summary>
    /// True when two folder names are the same release wrapped twice. Deliberately exact, not a similarity score: the
    /// names must agree on every token, except that a season marker may differ when the destination already names that
    /// season, and they must share at least one token that is not encode jargon. A false negative leaves a folder
    /// nested; a false positive merges two releases, so the asymmetry is intentional.
    /// </summary>
    public static bool IsSameRelease(string a, string b, IReadOnlySet<string>? destKeys = null)
    {
        var ta = ReleaseTokens(a);
        var tb = ReleaseTokens(b);
        if (ta.Count == 0 || tb.Count == 0) return false;

        var sharedMeaningful = ta.Count(t => tb.Contains(t) && !TechnicalToken.IsMatch(t));
        if (sharedMeaningful == 0) return false;

        bool Allowed(string t)
        {
            var season = SeasonMarker.Match(t);
            if (!season.Success) return false;
            // Only droppable if the destination already says which season this is.
            return destKeys?.Contains("season " + Num(season.Groups[1].Value)) ?? false;
        }

        foreach (var t in ta) if (!tb.Contains(t) && !Allowed(t)) return false;
        foreach (var t in tb) if (!ta.Contains(t) && !Allowed(t)) return false;
        return true;
    }

    /// <summary>
    /// The destination components a container folder may repeat. Only the tail: a folder named <c>Downloads</c> says
    /// nothing about the torrent just because the library lives under a folder of that name.
    /// </summary>
    public static HashSet<string> DestinationKeys(string? destPath) =>
        Segments(destPath).TakeLast(3).Select(FolderKey).ToHashSet(StringComparer.Ordinal);

    /// <summary>Decides whether one container level may be dropped. This function <em>is</em> the layout agreement.</summary>
    /// <param name="name">the folder in question</param>
    /// <param name="depth">0 for the torrent's own container root</param>
    /// <param name="destKeys">from <see cref="DestinationKeys"/></param>
    /// <param name="ancestors">folders already dropped, outermost first</param>
    public static bool MayDropFolder(string name, int depth, IReadOnlySet<string> destKeys, IReadOnlyList<string> ancestors)
    {
        if (ProtectedFolder.IsMatch(name)) return false;

        var repeatsDestination = destKeys.Contains(FolderKey(name));

        // A structural folder is the only thing separating two discs or two seasons.
        if (StructuralFolder.IsMatch(name)) return repeatsDestination;

        // Depth 0 is the torrent's container root, redundant by construction — qBittorrent's
        // contentLayout=NoSubfolder, so a download moved between the two clients still finds its files.
        if (depth == 0) return true;

        // Deeper down, absence of meaning is not evidence. Require proof.
        return repeatsDestination || ancestors.Any(prior => IsSameRelease(prior, name, destKeys));
    }

    /// <summary>True when the destination path itself already names one season.</summary>
    public static bool DestinationNamesSeason(string? destPath)
    {
        var segs = Segments(destPath);
        return segs.Count > 0 && SeasonSegment.IsMatch(segs[^1]);
    }

    /// <summary>
    /// Renames a release folder that names exactly one season into <c>Season NN</c>. Never a protected or structural
    /// folder, the name must carry an encode token (so a title containing <c>S2</c> is not mistaken for one), and it
    /// must resolve to exactly one season number. Null leaves the folder untouched, which is always the safe answer.
    /// </summary>
    public static string? SeasonFolderRename(string name)
    {
        if (ProtectedFolder.IsMatch(name)) return null;
        if (StructuralFolder.IsMatch(name)) return null;

        var seasons = SeasonMention.Matches(name)
            .Select(m => int.Parse(m.Groups[1].Value, CultureInfo.InvariantCulture)).ToHashSet();
        if (seasons.Count != 1) return null;

        var tokens = NonAlnum.Split(name.ToLowerInvariant()).Where(t => t.Length > 0);
        if (!tokens.Any(t => TechnicalToken.IsMatch(t))) return null;

        return "Season " + seasons.First().ToString("00", CultureInfo.InvariantCulture);
    }

    /// <summary>
    /// The key two paths collide on <em>as files on disk</em>. Reserved characters are stripped from the basename (the
    /// store does it on every platform), and Windows also folds case and ignores trailing dots and spaces.
    /// </summary>
    public static string PhysicalKey(string relativePath, bool? windows = null)
    {
        var win = windows ?? OperatingSystem.IsWindows();
        var parts = Segments(relativePath);
        var cleaned = parts.Select((seg, i) =>
        {
            var s = seg;
            // Only the basename is sanitised by the store; Windows still folds directories' case and trailing punctuation.
            if (i == parts.Count - 1) s = ReservedFilename.Replace(s, "");
            if (win) s = TrailingDotsSpaces.Replace(s, "");
            return s;
        });
        var joined = string.Join('/', cleaned);
        return win ? joined.ToLowerInvariant() : joined;
    }

    /// <summary>
    /// True when <paramref name="child"/> is at or below <paramref name="root"/>, resolving <c>..</c> first. Compares
    /// segments, not a string prefix: a folder called <c>..stfolder</c> starts with <c>..</c> without escaping anything.
    /// </summary>
    public static bool IsInside(string root, string child)
    {
        var rel = Path.GetRelativePath(Path.GetFullPath(root), Path.GetFullPath(child));
        if (rel == ".") return true;
        if (Path.IsPathRooted(rel)) return false;
        return rel != ".." && !rel.StartsWith(".." + Path.DirectorySeparatorChar, StringComparison.Ordinal);
    }
}
